### PR TITLE
Added the API and infrastructure to show a progress bar (CRAFT-99).

### PR DESCRIPTION
**Pull request by facundobatista**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_cli/messages.py [hunk 1]
```diff
@@ -134,14 +137,53 @@ def _write_line(self, message: _MessageInfo) -> None:
...
+        # terminal size minus the text and numerical progress, and 5 (the cursor at the end,
+        # two spaces before and after the bar, and two surrounding brackets)
+        terminal_width = _get_terminal_width()
+        bar_width = terminal_width - len(message.text) - len(numerical_progress) - 5
+
+        # only show the bar with progress if there is enough space, otherwise just the
+        # message (truncated, if needed)
+        if bar_width > 0:
+            completed_width = math.floor(bar_width * min(bar_percentage, 100))
+            completed_bar = "#" * completed_width
```
Comment (cmatsuoka) [inline craft_cli/messages.py:165]: We could consider changing to a block character at some point in the future.

File: tests/unit/test_messages_emitter.py [hunk 2]
```diff
@@ -274,6 +274,46 @@ def test_progress_in_verboseish_modes(get_initiated_emitter, mode):
...
+        call().show(sys.stderr, "some text", ephemeral=True),
+    ]
+    assert progresser.total == 5000
+    assert progresser.text == "some text"
+    assert progresser.stream == sys.stderr
+    assert progresser.delta is True
+
+
+def test_progressbar_with_delta_false(get_initiated_emitter):
+    """Init _Progresser with delta=Falsee."""
```
Comment (cmatsuoka) [inline tests/unit/test_messages_emitter.py:300]: Typo, extra "e" in `False`.

File: tests/unit/test_messages_helpers.py [hunk 3]
```diff
@@ -127,3 +129,46 @@ def test_getlogpath_ignore_other_files(test_log_dir, monkeypatch):
...
+
+def test_progresser_absolute_mode():
+    """Just use _Progresser as a context manager in absolute mode."""
+    stream = sys.stdout
+    text = "test text"
+    total = 123
+    fake_printer = MagicMock()
+    with _Progresser(fake_printer, total, text, stream, delta=False) as progresser:
+        progresser.advance(20)
+        progresser.advance(30)
```
Comment (cmatsuoka) [inline tests/unit/test_messages_helpers.py:145]: I think we could also test for out of range values, and also floating point numbers since we allow floats.

File: tests/unit/test_messages_helpers.py [hunk 4]
```diff
@@ -127,3 +129,46 @@ def test_getlogpath_ignore_other_files(test_log_dir, monkeypatch):
...
+
+def test_progresser_delta_mode():
+    """Just use _Progresser as a context manager in delta mode."""
+    stream = sys.stdout
+    text = "test text"
+    total = 123
+    fake_printer = MagicMock()
+    with _Progresser(fake_printer, total, text, stream, delta=True) as progresser:
+        progresser.advance(20)
+        progresser.advance(30)
```
Comment (cmatsuoka) [inline tests/unit/test_messages_helpers.py:161]: We could also test for delta values that bring the accumulated total to be out of range, and also using floating point numbers since we allow floats.

File: tests/unit/test_messages_printer.py [hunk 5]
```diff
@@ -232,6 +232,170 @@ def test_writeline_having_previous_message_ephemeral(capsys, monkeypatch, log_fi
...
+    """The indicated progress exceeds the total."""
+    monkeypatch.setattr(messages, "_get_terminal_width", lambda: 40)
+    printer = _Printer(log_filepath)
+
+    msg = _MessageInfo(sys.stdout, "test text", bar_progress=120, bar_total=100)
+    printer._write_bar(msg)
+
+    out, _ = capsys.readouterr()
+    assert len(out) == 39
+    assert out == "test text [###################] 120/100"
```
Comment (cmatsuoka) [inline tests/unit/test_messages_printer.py:292]: Not related to this test, but will the output look correct if `bar_progress` is set to something like `1000/100`, and then back to `120/100`?

File: craft_cli/messages.py [hunk 6]
```diff
@@ -134,14 +137,53 @@ def _write_line(self, message: _MessageInfo) -> None:
...
+        elif self.prv_msg.ephemeral:
+            # the last one was ephemeral, overwrite it
+            maybe_cr = "\r"
+        else:
+            # complete the previous line, leaving that message ok
+            maybe_cr = ""
+            print(flush=True, file=self.prv_msg.stream)
+
+        numerical_progress = f"{message.bar_progress}/{message.bar_total}"
+        bar_percentage = min(message.bar_progress / message.bar_total, 1)  # type: ignore
```
Comment (cmatsuoka) [inline craft_cli/messages.py:154]: Should we clamp out of range progress values in the model or only for display? What is the meaning of negative progress values, if `_Progresser.advance()` is called with a negative number in the absolute case, or if it falls below 0 in the delta case?

File: craft_cli/messages.py (hunk 1 above)
Comment (facundobatista) [reply craft_cli/messages.py:165]: Yes. @sergiusens already mentioned it. I just didn't know which character to use. Do you have one that we know to be in the different fonts that are used in terminals both in Linux and Windows?

File: tests/unit/test_messages_emitter.py (hunk 2 above)
Comment (facundobatista) [reply tests/unit/test_messages_emitter.py:300]: Fixed, thanks!

File: tests/unit/test_messages_helpers.py (hunk 3 above)
Comment (facundobatista) [reply tests/unit/test_messages_helpers.py:145]: Good idea about floating point. Regarding "out of range", we're not doing anything special with them (the code that actually *shows* the progress deals with that)... what kind of test are you suggesting? Just that it doesn't crash?

File: tests/unit/test_messages_helpers.py (hunk 4 above)
Comment (facundobatista) [reply tests/unit/test_messages_helpers.py:161]: Same question above.

File: tests/unit/test_messages_printer.py (hunk 5 above)
Comment (facundobatista) [reply tests/unit/test_messages_printer.py:292]: Yes. Each time the line is drawn, it uses all the terminal's width.

File: craft_cli/messages.py (hunk 6 above)
Comment (facundobatista) [reply craft_cli/messages.py:154]: We allow out of range values. The bar itself for sure won't reflect it (it will show just full), but the values shown indicate that. My idea here is that we will not "hide" a detail like that in the lib. 

If the app has a bug or something that gets into this case, the user and us would see it, and fix it. If the lib hides it, the progress would just get "stuck at 100%", while the app continues working.

On the other hand, negative values should not be allowed, as they are just wrong. I changed `_Progresser` to raise an error in that case.

File: craft_cli/messages.py (hunk 1 above)
Comment (sergiusens) [reply craft_cli/messages.py:165]: This should work https://github.com/verigak/progress/blob/master/progress/bar.py#L52, from this post, looks like the ones mentioned in this table https://mike42.me/blog/2018-06-make-better-cli-progress-bars-with-unicode-block-characters would work too

**Review by sergiusens — APPROVED** [review]

File: craft_cli/messages.py (hunk 1 above)
Comment (facundobatista) [reply craft_cli/messages.py:165]: I'd use the one that you mentioned first. It's unicode name is 'FULL BLOCK'... looks simple enough, it should be in all fonts (who know!).

We can leave "sub char resolution" for the future, IMO. I just opened #9 to not forget about it.

Comment (facundobatista) [reply craft_cli/messages.py]: Pushed the version using the FULL BLOCK.

File: tests/unit/test_messages_helpers.py (hunk 3 above)
Comment (cmatsuoka) [reply tests/unit/test_messages_helpers.py]: Yes, to make sure it doesn't crash.

**Review by cmatsuoka — APPROVED** [review]
Thanks for the clarification.
